text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,299 @@
+1. The lifecycle will run the detect phase for all stackpacks defined in the builder.
+1. The lifecycle will execute the stack buildpack build phase for all passing stackpack(s). If no stack buildpacks pass detect, the build will continue the build phase as normal (running userspace buildpacks).
+1. After, during, or before the lifecycle's build phase, the lifecycle will begin a new phase called _extend_.
+1. During the extend phase, stack buildpacks that passed detection will run against the run images accordingly (see details below).
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+ A stack buildpack (a.k.a. stackpack) is a special case of buildpack that has the following properties:
+
+* Is run as the `root` user
+* Configured with `privileged = true` in the `buildpack.toml`
+* Can only create one layer
+* Does not have access to the application source code
+* Is run before all regular buildpacks
+* Is run against both the build and run images
+* Is distributed with the stack run and/or build image
+* May not create layers using the `<layers>` directory
+
+The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.
+
+For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase excluding the a few specific directories and files.
+
+Additionally, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+A stack can provide stackpacks by including them in the `/cnb/stack/buildpacks` directory. By default, an implicit order will be created for all stackpacks included in a stack. The order can be overridden in the `builder.toml` with the following configuration:
+
+```
+[[stack.order]]
+  [[stack.order.group]]
+    id = "<stackpack id>"
+```
+
+A stackpack will only execute if it passes detection. When the stackpack is executed, its detect and build scripts use the same parameters as the regular buildpacks. But the arguments may differ. For example, the first positional argument to the `bin/build` (the `<layers>` directory) MUST NOT be a standard layers directory adhering to the [Buildpacks Build specification](https://github.com/buildpacks/spec/blob/main/buildpack.md#build).
+
+The stackpack's snapshot layer may be modified by writing a `stack.toml` file. The `stack.toml` will define paths that will be restored even when the base image changes (ex. package indicies) and paths that will be excluded from the launch image (ex. `/var/cache`).
```

Comment:
> The `stack.toml` will define paths that will be restored even when the base image changes

I think we need a schema for `stack.toml` and a more thorough description of the behavior in this RFC.

What do we mean by "paths that will be restored even when the base image changes"? Here are the different ways I can interpret this:

1. We creating and caching a separate cache layer that is a subset of the snapshot layer and putting it in a directory or image cache? We restore this during the traditional "restore" phase.
    * In this case wouldn't we be restoring on every build just like a normal cache layer? "even when the base image changes" feel unnecessary. 
    * Should we actually be slicing "exactly one" layer into multiple layers so we can talk about their properties (launch/cache etc) in a straightforward way?
2. We potentially add a restore component to rebase? But that doesn't make sense b/c a normal rebase should work here and we haven't RFC'd or specified an hypothetical `upgrade` operation yet.

Reply:
this should be `launch.toml`. The schema is down below, but I can definitely add more detail like this in there too

Reply:
I think the intended behavior is that for stackpack builds in all three scenarios (build-time build, build-time run, and rebase run), globs of files may be marked as layers to be excluded the image when it is "used." Any excluded layer may also be marked as cached, so that those excluded layers are recovered when the image is "used." What "used" means is defined as:

1. **Build-time build:** A given layer is excluded at normal buildpack build-time, and recovered the next time the build image is extended with the stackpack.
2. **Build-time run:** A given layer is excluded from the final image, and recovered the next time the run image is extended with the stackpack (either rebase or rebuild).
2. **Rebase run:** A given layer is excluded from the rebased image, and recovered the next time the run image is extended with the stackpack (either rebase or rebuild).

Reply:
Addressed in 09316a1

Reply:
> Build-time build: A given layer is excluded at normal buildpack build-time, and recovered the next time the build image is extended with the stackpack.

What do we mean by "excluded at normal buildpack build-time" exactly? That is not possible without undoing the changes right?